### PR TITLE
wallet: Fix listwalletdir listing of migrated default wallets and generated backup files

### PR DESCRIPTION
**Pull request by achow101**
When the default wallet is migrated, we do not rename the wallet so we end up having a descriptor wallet with the empty string as its name and the wallet.dat file in the root of the walletdir. This is supposed to be an unsupported configuration and there is no other way to achieve this (other than file copying), but the wallet loading code does not disallow loading such wallets. However `listwalletdir` does not currently list the default wallet if it is sqlite. This is confusing to users, so change `listwalletdir` to include these wallets.

Additionally, the migration of the default wallet, and of any plain wallet files in the walletdir, produces a backup file in the walletdir itself. Since these backups are a BDB file, `listwalletdir` will detect them as being another wallet that we could open, but this is erroneous and could lead to confusion and potentially funds loss if both the backup and the migrated wallet are in use simultaneously. To reduce the likelihood of this issue, don't list these wallets in `listwalletdir`.

***

Possibly we could have more stringent checks on loading to resolve these issues, but I'm concerned that that will just confuse users and gratuitously break things that already worked.

Since the original intent was to disallow default wallets for sqlite/descriptors, a possible alternative would be to prevent people from loading such wallets and change migration to rename those wallets. However, given that this behavior with migrating default wallets has existed since default wallet migration was fixed, I think that making such a change would be confusing and break things for no good reason. Although perhaps we should still do the renaming.

For the backups, we could also change loading to refuse to load any wallet named with `.bak` (or `.legacy.bak`) as such wallets can still be loaded by giving the path to them directly, which some users may do to "restore" the backup. However restricting what can be loaded based on filename seems a little heavyhanded. It wouldn't be funds loss though since the correct way to restore the backup is with `restorewallet`.

### CODE REVIEW

**Comment by Eunovo** [comment]
Tested ACK https://github.com/bitcoin/bitcoin/pull/30265/commits/0ce7c6294bf58142de55372d6448416a73f4caa5
I ran the `wallet_migration` test locally and also prepared a wallets directory with a default wallet file, a backup file and a legacy wallet file for manual testing. Running `listwalletdir` on master returned the following output
```
{
  "wallets": [
    {
      "name": "boblegacy_1718798800.legacy.bak"
    },
    {
      "name": "boblegacy"
    }
  ]
}
```
And `listwalletdir` on https://github.com/bitcoin/bitcoin/pull/30265/commits/0ce7c6294bf58142de55372d6448416a73f4caa5 returned this
```
{
  "wallets": [
    {
      "name": "boblegacy"
    },
    {
      "name": ""
    }
  ]
}
```
The no-name wallet is displayed and the backup wallet file is hidden as expected.

File: test/functional/wallet_migration.py [hunk 1]
```diff
@@ -538,10 +538,16 @@ def test_default_wallet(self):
...
+
         # Check backup existence and its non-empty wallet filename
-        backup_path = self.nodes[0].wallets_path / f'default_wallet_{curr_time}.legacy.bak'
+        backup_filename = f"default_wallet_{curr_time}.legacy.bak"
+        backup_path = self.nodes[0].wallets_path / backup_filename
         assert backup_path.exists()
         assert_equal(str(backup_path), res['backup_path'])
+        assert {"name": backup_filename} not in walletdir_list["wallets"]
+
+        wallet.setmocktime(0)
```
Comment (furszy) [inline test/functional/wallet_migration.py:551]: this line does not seems to be related to a700b303fcba7915f.

Comment (achow101) [reply test/functional/wallet_migration.py]: Removed

**Review by furszy — COMMENTED** [review]
Code ACK  6b2dcba076

**Comment by fjahr** [comment]
Code review ACK 6b2dcba07670f04f32c0dc3a2c86fd805c85f12d

Also verified manually that the added tests do cover the change of behavior.

**Review by glozow — COMMENTED** [review]
ACK 6b2dcba07670f04f32c0dc3a2c86fd805c85f12d

Tested by creating legacy default and non-default wallets, then migrating, then using `listwalletdir`. I can see the default wallet is listed as a .bak on master, then disappears on 3ddbdd1815c676a88345b3b0e55a551d2a569e28, then is listed with its name on 6b2dcba07670f04f32c0dc3a2c86fd805c85f12d.

**Pull request merged**